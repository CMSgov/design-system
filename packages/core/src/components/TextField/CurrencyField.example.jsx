import React from 'react';
import ReactDOM from 'react-dom';
import TextField from './TextField';

ReactDOM.render(
  <TextField
    ariaLabel="Enter monthly income amount in dollars."
    label="Monthly income"
    mask="currency"
    name="currency_example"
    value="2500"
  />,
  document.getElementById('js-example')
);
