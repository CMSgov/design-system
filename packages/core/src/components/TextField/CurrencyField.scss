@import '@cmsgov/design-system-support/src/settings/index';

/*
Currency field

A currency field is an enhanced input field that provides visual and non-visual
cues to a user to enter an amount of money in a specified currency.

Markup:
<label class="ds-c-label" for="currency_field">
  Monthly income
</label>
<div class="ds-c-field-mask ds-c-field-mask--currency">
  <div class="ds-c-field__before ds-c-field__before--currency">$</div>
  <input type="text" aria-label="Enter amount in dollars" class="ds-c-field ds-c-field--currency" id="currency_field" name="currency_example">
</div>

Style guide: components.currency-field
*/

/*
`<TextField mask="currency">`

Passing a `mask="currency"` prop into the `TextField` component will
enable formatting to occur when the field is blurred. To "unmask" the
value, you can import the `unmaskValue` method.

@react-component TextField

@react-example CurrencyField

Style guide: components.currency-field.react
*/
.ds-c-field__before--currency {
  // Increase currency size so it serves more as an icon and
  // better aligns with the input value text
  font-size: $base-font-size + 2;
  font-weight: $font-bold;
  // Adjust to better align icon with value's baseline
  padding-top: 1px;
}

.ds-c-field--currency {
  // Normal padding + space for the icon + normal padding
  padding-left: $input-padding + 4px + $input-padding;
}

/*
---

## When to use

- Use the Currency input component anytime users should enter an amount of money in a particular currency, like dollars.

**[View the "Text field" guidance for additional guidance and best practices.]({{root}}/patterns/text-field#guidance)**

## Accessibility

- Ensure the field includes an `aria-label` describing the expected currency for users using assistive devices.

## Related patterns

- [Text field]({{root}}/components/text-field/)

## Learn more

- [Form Guidelines]({{root}}/guidelines/forms/)
- [GOV.UK - Currency input testing and research](https://github.com/alphagov/govuk-design-system/wiki/Currency-input-testing-and-research)

Style guide: components.currency-field.guidance
*/
