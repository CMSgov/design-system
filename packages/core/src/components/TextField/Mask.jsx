/*
Masked field

A masked field is an enhanced input field that provides visual and non-visual
cues to a user about the expected value.

Style guide: components.masked-field
*/
import 'core-js/fn/array/includes';
import PropTypes from 'prop-types';
import React from 'react';
import { polyfill } from 'react-lifecycles-compat';

// Deliminate chunks of integers
const deliminatedMaskRegex = {
  phone: /(\d{3})(\d{1,3})?(\d+)?/,
  ssn: /([*\d]{3})([*\d]{1,2})?([*\d]+)?/,
  zip: /(\d{5})(\d*)/
};

/**
 * Split value into groups and insert a hyphen deliminator between each
 * @param {String} value
 * @param {RegExp} rx - Regular expression with capturing groups
 * @returns {String}
 */
function deliminateRegexGroups(value, rx) {
  const matches = toDigitsAndAsterisks(value).match(rx);

  if (matches && matches.length > 1) {
    value = matches
      .slice(1)
      .filter(a => !!a) // remove undefined groups
      .join('-');
  }

  return value;
}

/**
 * Format a string using fixed-point notation, similar to Number.prototype.toFixed
 * though a decimal is only fixed if the string included a decimal already
 * @param {String} value - A stringified number (i.e. "1234")
 * @param {Number} digits - The number of digits to appear after the decimal point
 * @returns {String}
 */
function stringWithFixedDigits(value, digits = 2) {
  const decimalRegex = /\.[\d]+$/;

  // Check for existing decimal
  const decimal = value.match(decimalRegex);

  if (decimal) {
    const fixedDecimal = parseFloat(decimal)
      .toFixed(digits)
      .match(decimalRegex)[0];

    return value.replace(decimal, fixedDecimal);
  }

  return value;
}

/**
 * Remove everything that isn't a digit or asterisk
 * @param {String} value
 * @returns {String}
 */
function toDigitsAndAsterisks(value) {
  return value.replace(/[^\d*]/g, '');
}

/**
 * Remove all non-digits
 * @param {String} value
 * @returns {String}
 */
function toDigits(value) {
  return value.replace(/[^\d]/g, '');
}

/**
 * Convert string into a number (positive or negative float or integer)
 * @param {String} value
 * @returns {Number}
 */
function toNumber(value) {
  if (typeof value !== 'string') return value;
  if (!value.match(/\d/)) return undefined;

  const sign = value.charAt(0) === '-' ? -1 : 1;
  const parts = value.split('.');
  // This assumes if the user adds a "." it should be a float. If we want it to
  // evaluate as an integer if there are no digits beyond the decimal, then we
  // can change it.
  const hasDecimal = parts[1] !== undefined;
  if (hasDecimal) {
    const a = toDigits(parts[0]);
    const b = toDigits(parts[1]);
    return sign * parseFloat(`${a}.${b}`);
  } else {
    return sign * parseInt(toDigits(parts[0]));
  }
}

/**
 * Returns the value with additional masking characters
 * @param {String} value
 * @returns {String}
 */
function maskValue(value = '', mask) {
  if (value && typeof value === 'string') {
    value = value.trim();

    if (mask === 'currency') {
      // Format number with commas. If the number includes a decimal,
      // ensure it includes two decimal points
      const number = toNumber(value);
      if (number === undefined) {
        value = '';
      } else {
        value = stringWithFixedDigits(number.toLocaleString('en-US'));
      }
    } else if (Object.keys(deliminatedMaskRegex).includes(mask)) {
      value = deliminateRegexGroups(value, deliminatedMaskRegex[mask]);
    }
  }

  return value;
}

/*
`<TextField mask={...}>`

Passing a `mask` prop into the `TextField` component with a valid value will
enable formatting to occur when the field is blurred. To "unmask" the
value, you can import and call the `unmaskValue` method.

@react-component TextField

@react-example Mask

Style guide: components.masked-field.react
*/

/**
 * A Mask component renders a controlled input field. When the
 * field is blurred, it applies formatting to improve the readability
 * of the value.
 */
class _Mask extends React.PureComponent {
  static getDerivedStateFromProps(props, state) {
    const fieldProps = React.Children.only(props.children).props;
    const isControlled = fieldProps.value !== undefined;
    if (isControlled) {
      const { mask } = props;
      if (unmask(fieldProps.value, mask) !== unmask(state.value, mask)) {
        return {
          value: maskValue(fieldProps.value || '', mask)
        };
      }
    }
    return null;
  }

  constructor(props) {
    super(props);

    const field = this.field();
    const initialValue = field.props.value || field.props.defaultValue;
    // console.log('initial value', initialValue, maskValue(initialValue, props.mask), props.mask)

    this.state = {
      value: maskValue(initialValue, props.mask)
    };
  }

  componentDidUpdate() {
    if (this.debouncedOnBlurEvent) {
      this.field().props.onBlur(this.debouncedOnBlurEvent);
      this.debouncedOnBlurEvent = null;
    }
  }

  /**
   * Get the child text field. Called as a method so that
   * updates to the field cause the mask to re-render
   * @returns {React.ReactElement} Child TextField
   */
  field() {
    return React.Children.only(this.props.children);
  }

  /**
   * To avoid a jarring experience for screen readers, we only
   * add/remove characters after the field has been blurred,
   * rather than when the user is typing in the field
   * @param {Object} evt
   * @param {React.Element} field - Child TextField
   */
  handleBlur(evt, field) {
    const value = maskValue(evt.target.value, this.props.mask);

    // We only debounce the onBlur when we know for sure that
    // this component will re-render (AKA when the value changes)
    // and when an onBlur callback is present
    const debounce =
      value !== this.state.value && typeof field.props.onBlur === 'function';

    if (debounce) {
      // We need to retain a reference to the event after the callback
      // has been called. We pass this onto the consuming app's onBlur
      // only after the value has been manipulated â€“ this way, the
      // value returned by event.target.value is the value after masking
      evt.persist();
      this.debouncedOnBlurEvent = evt;
    }

    this.setState({
      value
    });

    if (!debounce && typeof field.props.onBlur === 'function') {
      // If we didn't debounce the onBlur event, then we need to
      // call the onBlur callback from here
      field.props.onBlur(evt);
    }
  }

  /**
   * @param {Object} evt
   * @param {React.Element} field - Child TextField
   */
  handleChange(evt, field) {
    this.setState({ value: evt.target.value });

    if (typeof field.props.onChange === 'function') {
      field.props.onChange(evt);
    }
  }

  render() {
    const field = this.field();

    return React.cloneElement(field, {
      defaultValue: undefined,
      onBlur: evt => this.handleBlur(evt, field),
      onChange: evt => this.handleChange(evt, field),
      value: this.state.value
    });
  }
}

_Mask.propTypes = {
  /** Pass the input as the child */
  children: PropTypes.node.isRequired,
  mask: PropTypes.string.isRequired
};

/**
 * Remove mask characters from value
 * @param {String} value
 * @param {String} mask
 * @returns {String}
 */
export function unmask(value, mask) {
  if (!value || typeof value !== 'string') return value;
  const rawValue = value;
  value = value.trim();

  if (mask === 'currency') {
    // Preserve only digits, decimal point, or negative symbol
    const matches = value.match(/^-|[\d.]/g);
    if (matches) {
      value = matches.join('');
    } else {
      value = '';
    }
  } else if (Object.keys(deliminatedMaskRegex).includes(mask)) {
    // Remove the deliminators and revert to single ungrouped string
    value = toDigitsAndAsterisks(value);
  } else {
    return rawValue;
  }

  return value;
}

const Mask = polyfill(_Mask);

export { Mask };
export default Mask;
