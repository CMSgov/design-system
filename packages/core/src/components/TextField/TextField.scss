@import '@cmsgov/design-system-support/src/settings/index';

.ds-c-field {
  appearance: none;
  border: $input-border-width solid $input-border-color;
  border-radius: $input-border-radius;
  box-sizing: border-box; // ensure padding doesn't affect width
  color: $color-base;
  display: block;
  margin: $spacer-half 0;
  max-width: $input-max-width;
  outline: none;
  padding: $input-padding;
  width: 100%;

  &:disabled {
    background-color: $color-gray-lighter;
    border-color: $color-gray-light;
  }

  &:focus {
    border-radius: 0;
    box-shadow: $focus-shadow;
    outline: 3px solid $focus-color;
    outline-offset: 0;
  }
}

.ds-c-field--small {
  max-width: $input-small-width;
}

.ds-c-field--medium {
  max-width: $input-medium-width;
}

.ds-c-field,
.ds-c-field-mask {
  font-family: $font-sans;
  font-size: $base-font-size;
  line-height: $input-line-height;
}

/*
Inverse theme
*/

.ds-c-field--inverse {
  border-color: $input-border-color-inverse;

  &:focus {
    box-shadow: $focus-shadow-inverse;
  }
}

// State modifiers and message
// ==============================
.ds-c-field--error {
  border: 3px solid $color-error;

  &:focus {
    border-color: $color-base;
    box-shadow: none;
  }
}

.ds-c-field--success {
  border: 3px solid $color-success-light;

  &:focus {
    border-color: $color-base;
    box-shadow: none;
  }
}
