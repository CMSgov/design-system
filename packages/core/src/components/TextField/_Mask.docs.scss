/*
Masked field

A masked field is an enhanced input field that provides visual and non-visual
cues to a user about the expected value.

Style guide: components.masked-field
*/

/*
`<TextField mask={...}>`

Passing a `mask` prop into the `TextField` component with a valid value will
enable formatting to occur when the field is blurred. To "unmask" the
value, you can import and call the `unmaskValue` method.

@react-component TextField

@react-example Mask

Style guide: components.masked-field.react
*/
