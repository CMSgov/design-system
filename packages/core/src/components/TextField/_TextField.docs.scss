/*
Text field

@uswds https://standards.usa.gov/components/form-controls/#text-input

Text fields allow users to enter any combination of letters, numbers, or symbols of their choosing (unless otherwise restricted). Text fields can span single or multiple lines.

Markup: textfield.example.html

Style guide: components.text-field
*/

/*
Inverse theme

Markup: textfield-inverse.example.html

Style guide: components.text-field.inverse
*/

/*
`<TextField>`

@react-component TextField

Style guide: components.text-field.react
*/

/*
---

### When to use

- If you can’t reasonably predict a user’s answer to a prompt and there might be wide variability in users’ answers.
- When using another type of input will make answering more difficult. For example, birthdays and other known dates are easier to type in than they are to select from a calendar picker.
- When users want to paste in a response.

### When to consider alternatives

- When users are choosing from a specific set of options. Consider [checkboxes, radio buttons](/components/choice), or a [select menu](/components/select) in these cases.

### Usage

- Don't use placeholder text in form fields. Use hint text instead, if you need to provide contextual information. Placeholder text disappears after a user types a value, therefore users will no longer have that text available when they need to review their entries. People who have cognitive or visual disabilities have additional problems with placeholder text.
- The length of the text field provides a hint to users as to how much text to write. Do not require users to write paragraphs of text into a single-line `input` box; use a `textarea` instead.
- Text fields are among the easiest type of input for desktop users but are more difficult for mobile users.
- Only show error messages or styling after a user has interacted with a particular field.

**[View the "Forms" guidelines for additional guidance and best practices.](/guidelines/forms/)**

### Accessibility

- Group each set of thematically related controls in a `fieldset` element. Use the `legend` element to offer a label within each `fieldset`. The `fieldset` and `legend` elements make it easier for screen reader users to navigate the form.
- Keep your form blocks in a vertical pattern. This approach is ideal, from an accessibility standpoint, because of limited vision that makes it hard to scan from left to right.

### Customization

The following Sass variables can be overridden to theme a field:

- `$color-focus`
- `$color-focus-inverse`
- `$focus-shadow`
- `$focus-shadow-inverse`
- `$input-border-color`
- `$input-border-color-inverse`
- `$input-border-radius`
- `$input-line-height`
- `$input-border-width`
- `$input-padding`

### Related patterns

- [Masked field](/components/masked-field/)
- [Date field](/components/date-field/)

### Learn more

- [Form Guidelines](/guidelines/forms/)
- ["Placeholders in Form Fields Are Harmful"](https://www.nngroup.com/articles/form-design-placeholders/)
- [Asking for a date of birth](https://designnotes.blog.gov.uk/2013/12/05/asking-for-a-date-of-birth/)
- [GOV.UK text boxes discussion](https://paper.dropbox.com/doc/Text-boxes-8NLlgz9tjR8OzTz2N9wE3)
- [Four steps for choosing form elements on the Web (PDF)](http://www.formsthatwork.com/files/Articles/dropdown.pdf)

Style guide: components.text-field.guidance
*/
