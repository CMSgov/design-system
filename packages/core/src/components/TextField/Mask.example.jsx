import TextField, { unmaskValue } from './TextField';
import React from 'react';
import ReactDOM from 'react-dom';
// import { unmaskValue } from '@cmsgov/design-system-core';

function handleBlur(evt, mask) {
  console.log('Unmasked value:', unmaskValue(evt.target.value, mask));
}

const Example = () => {
  return (
    <div>
      <TextField
        ariaLabel="Enter monthly income amount in dollars."
        label="Currency"
        mask="currency"
        name="currency_example"
        onBlur={evt => handleBlur(evt, 'currency')}
        value="2500"
      />

      <TextField
        label="Phone number"
        mask="phone"
        name="phone_example"
        onBlur={evt => handleBlur(evt, 'phone')}
        type="tel"
        value="1234567890"
      />

      <TextField
        label="Social security number (SSN)"
        mask="ssn"
        name="ssn_example"
        onBlur={evt => handleBlur(evt, 'ssn')}
        value="123456789"
      />

      <TextField
        label="Zip code"
        mask="zip"
        name="zip_example"
        onBlur={evt => handleBlur(evt, 'zip')}
        value="123456789"
      />
    </div>
  );
};

ReactDOM.render(<Example />, document.getElementById('js-example'));
