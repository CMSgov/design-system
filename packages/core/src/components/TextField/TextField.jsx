import FormLabel from '../FormLabel/FormLabel';
import Mask from './Mask';
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import uniqueId from 'lodash.uniqueid';

export { unmaskValue } from './Mask';

export class TextField extends React.PureComponent {
  constructor(props) {
    super(props);
    this.id = props.id || uniqueId('textfield_');
    this.labelId = props.labelId || uniqueId('textfield_label_');

    if (process.env.NODE_ENV !== 'production') {
      if (props.fieldRef) {
        console.warn(
          `[Deprecated]: Please remove the 'fieldRef' prop in <TextField>, use 'inputRef' instead. This prop has been renamed and will be removed in a future release.`
        );
      }
      if (props.type === 'number') {
        console.warn(
          `Please use the 'numeric' prop instead of 'type="number"' unless your user research suggests otherwise.`
        );
      }
    }
  }

  componentDidMount() {
    if (this.props.focusTrigger) {
      this.focusRef && this.focusRef.focus();
    }
  }

  ariaLabel() {
    if (this.props.ariaLabel) {
      return this.props.ariaLabel;
    } else if (this.props.mask === 'currency') {
      return `${this.props.label}. Enter amount in dollars.`;
    }
  }

  /**
   * @param {React.Component} field
   * @returns {React.Component} The input field, optionally including mask
   *  markup if a mask is present
   */
  renderFieldAndMask(field) {
    return this.props.mask ? <Mask mask={this.props.mask}>{field}</Mask> : field;
  }

  render() {
    const {
      ariaLabel,
      className,
      errorMessage,
      fieldClassName,
      fieldRef,
      focusTrigger,
      hint,
      id,
      inversed,
      inputRef,
      label,
      labelClassName,
      labelId,
      mask,
      multiline,
      numeric,
      requirementLabel,
      rows,
      size,
      type,
      pattern,
      ...fieldProps
    } = this.props;
    const FieldComponent = multiline ? 'textarea' : 'input';
    const _rows = multiline && rows ? rows : undefined;

    const classes = classNames(
      'ds-u-clearfix', // fixes issue where the label's margin is collapsed
      className
    );

    const fieldClasses = classNames(
      'ds-c-field',
      mask && `ds-c-field--${mask}`,
      {
        'ds-c-field--error': typeof errorMessage === 'string',
        'ds-c-field--inverse': inversed
      },
      fieldClassName,
      size && `ds-c-field--${size}`
    );

    let inputType = type;
    if (numeric) {
      inputType = 'text';
    } else if (multiline) {
      inputType = undefined;
    }

    const field = (
      <FieldComponent
        aria-label={this.ariaLabel()}
        className={fieldClasses}
        id={this.id}
        /* eslint-disable no-return-assign */
        ref={ref => {
          if (focusTrigger) {
            this.focusRef = ref;
          } else {
            if (inputRef) {
              inputRef(ref);
            }
            if (fieldRef) {
              fieldRef(ref);
            }
          }
        }}
        /* eslint-enable no-return-assign */
        rows={_rows}
        inputMode={numeric ? 'numeric' : undefined}
        pattern={numeric && !pattern ? '[0-9]*' : pattern}
        type={inputType}
        {...fieldProps}
      />
    );

    return (
      <div className={classes}>
        <FormLabel
          className={labelClassName}
          errorMessage={errorMessage}
          fieldId={this.id}
          hint={hint}
          id={this.labelId}
          requirementLabel={requirementLabel}
          inversed={inversed}
        >
          {label}
        </FormLabel>
        {this.renderFieldAndMask(field, mask)}
      </div>
    );
  }
}

TextField.defaultProps = {
  type: 'text'
};

TextField.propTypes = {
  /**
   * Apply an `aria-label` to the text field to provide additional
   * context to assistive devices.
   */
  ariaLabel: PropTypes.string,
  /**
   * Additional classes to be added to the root `div` element
   */
  className: PropTypes.string,
  /**
   * Sets the initial value. Use this for an uncontrolled component; otherwise,
   * use the `value` property.
   */
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  disabled: PropTypes.bool,
  errorMessage: PropTypes.node,
  /**
   * Additional classes to be added to the field element
   */
  fieldClassName: PropTypes.string,
  /**
   * @hide-prop [Deprecated] Access a reference to the `input` or `textarea` element. Please use `inputRef` instead.
   */
  fieldRef: PropTypes.func,
  /**
   * Used to focus `input` on `componentDidMount()`
   */
  focusTrigger: PropTypes.bool,
  /**
   * Additional hint text to display
   */
  hint: PropTypes.node,
  /**
   * A unique `id` to be used on the text field.
   */
  id: PropTypes.string,
  /**
   * Access a reference to the `input` or `textarea` element
   */
  inputRef: PropTypes.func,
  /**
   * Text showing the requirement ("Required", "Optional", etc.). See [Required and Optional Fields]({{root}}/guidelines/forms/#required-and-optional-fields).
   */
  requirementLabel: PropTypes.node,
  /**
   * Applies the "inverse" UI theme
   */
  inversed: PropTypes.bool,
  /**
   * Label for the input
   */
  label: PropTypes.node.isRequired,
  /**
   * Additional classes to be added to the `FormLabel`.
   */
  labelClassName: PropTypes.string,
  /**
   * A unique `id` to be used on the label field.
   */
  labelId: PropTypes.string,
  /**
   * Apply formatting to the field that's unique to the value
   * you expect to be entered. Depending on the mask, the
   * field's appearance and functionality may be affected.
   */
  mask: PropTypes.oneOf(['currency', 'phone', 'ssn', 'zip']),
  /**
   * Whether or not the text field is a multiline text field
   */
  multiline: PropTypes.bool,
  name: PropTypes.string.isRequired,
  /**
   * Sets `inputMode`, `type`, and `pattern` to improve accessiblity and consistency for number fields. Use this prop instead of `type="number"`, see [here](https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/) for more information.
   */
  numeric: PropTypes.bool,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  /**
   * @hide-prop HTML `input` [pattern](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#htmlattrdefpattern).
   */
  pattern: PropTypes.string,
  /**
   * Optionally specify the number of visible text lines for the field. Only
   * applicable if this is a multiline field.
   */
  rows: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
   * Set the max-width of the input either to `'small'` or `'medium'`.
   */
  size: PropTypes.oneOf(['small', 'medium']),
  /**
   * HTML `input` [type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#<input>_types) attribute. If you are using `type=number` please use the numeric prop instead.
   */
  type: PropTypes.string,
  /**
   * Sets the input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `defaultValue`.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

export default TextField;
