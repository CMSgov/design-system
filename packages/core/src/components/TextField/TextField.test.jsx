import TextField, { unmaskValue } from './TextField';
import { mount, shallow } from 'enzyme';
import React from 'react';

function render(customProps = {}, deep = false) {
  const props = Object.assign(
    {
      label: 'Foo',
      name: 'spec-field'
    },
    customProps
  );
  const component = <TextField {...props} />;

  return {
    props: props,
    wrapper: deep ? mount(component) : shallow(component)
  };
}

describe('TextField', function() {
  it('is an input field', () => {
    const data = render();
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.is('input')).toBe(true);
    expect(field.prop('rows')).toBeUndefined();
    expect(field.prop('type')).toBe('text');
  });

  it('is a textarea', () => {
    const data = render({ multiline: true });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.is('textarea')).toBe(true);
    expect(field.prop('rows')).toBeUndefined();
    expect(field.prop('type')).toBeUndefined();
  });

  it('is a password field', () => {
    const data = render({ type: 'password' });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('type')).toBe('password');
  });

  it('is disabled', () => {
    const data = render({ disabled: true });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('disabled')).toBe(data.props.disabled);
  });

  it('has a defaultValue', () => {
    const data = render({ defaultValue: 'Yay' });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('defaultValue')).toBe(data.props.defaultValue);
    expect(field.prop('value')).toBeUndefined();
  });

  it('has a value', () => {
    const data = render({ value: 'Yay' });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('value')).toBe(data.props.value);
    expect(field.prop('defaultValue')).toBeUndefined();
  });

  it('shows 5 rows of text', () => {
    const data = render({
      multiline: true,
      rows: 5
    });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('rows')).toBe(data.props.rows);
  });

  it('has a unique id', () => {
    const fieldData1 = render();
    const fieldData2 = render();
    const field1 = fieldData1.wrapper.find('.ds-c-field').first();
    const label1 = fieldData1.wrapper.find('FormLabel').first();
    const field2 = fieldData2.wrapper.find('.ds-c-field').first();
    const idRegex = new RegExp('textfield_[0-9]+');

    expect(field1.prop('id')).toMatch(idRegex);
    expect(label1.prop('fieldId')).toBe(field1.prop('id'));
    expect(field1.prop('id')).not.toBe(field2.prop('id'));
  });

  it('has a label', () => {
    const data = render();
    const label = data.wrapper.find('FormLabel').first();

    expect(label.prop('children')).toBe(data.props.label);
  });

  it('has a hint', () => {
    const data = render({ hint: '123' });
    const label = data.wrapper.find('FormLabel').first();

    expect(label.prop('hint')).toBe(data.props.hint);
  });

  it('adds className to root element', () => {
    const data = render({ className: 'bar' });

    expect(data.wrapper.hasClass('bar')).toBe(true);
  });

  it('adds className to field', () => {
    const data = render({ fieldClassName: 'bar' });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.hasClass('ds-c-field')).toBe(true);
    expect(field.hasClass('bar')).toBe(true);
  });

  it('adds className to label', () => {
    const data = render({ labelClassName: 'bar' });

    expect(data.wrapper.find('FormLabel').hasClass('bar')).toBe(true);
  });

  it('adds size classes to input', () => {
    const mediumData = render({ size: 'medium' });
    const mediumField = mediumData.wrapper.find('.ds-c-field').first();
    const smallData = render({ size: 'small' });
    const smallField = smallData.wrapper.find('.ds-c-field').first();

    expect(mediumField.hasClass('ds-c-field--medium')).toBe(true);
    expect(smallField.hasClass('ds-c-field--small')).toBe(true);
  });

  it('adds min/max input attributes', () => {
    const data = render({
      max: 10,
      min: 1
    });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('max')).toBe(data.props.max);
    expect(field.prop('min')).toBe(data.props.min);
  });

  it('adds aria-label attribute', () => {
    const data = render({
      ariaLabel: 'Foo'
    });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('aria-label')).toBe(data.props.ariaLabel);
  });

  it('adds default aria-label for currency mask', () => {
    const data = render({
      mask: 'currency'
    });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('aria-label')).toBe(`${data.props.label}. Enter amount in dollars.`);
  });

  it('adds overrides default aria-label with defined prop', () => {
    const data = render({
      ariaLabel: 'Foo',
      mask: 'currency'
    });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('aria-label')).toBe(data.props.ariaLabel);
  });

  it('adds undocumented prop to input field', () => {
    const data = render({
      'data-foo': 'bar'
    });
    const field = data.wrapper.find('.ds-c-field').first();

    expect(field.prop('data-foo')).toBe(data.props['data-foo']);
  });

  it('returns reference to input field', () => {
    let ref;
    const data = render(
      {
        defaultValue: 'Yay',
        fieldRef: el => {
          ref = el;
        }
      },
      true
    );

    expect(ref.value).toBe(data.props.defaultValue);
  });

  it('focuses the input when focusTrigger is passed', () => {
    const data = render(
      {
        id: 'focus',
        focusTrigger: true
      },
      true
    );

    expect(data.wrapper.find('input').props().id).toEqual(document.activeElement.id);
  });

  describe('has error', () => {
    let data;

    beforeEach(() => {
      data = render({ errorMessage: 'Error' });
    });

    it('passes error to FormLabel', () => {
      expect(data.wrapper.find('FormLabel').prop('errorMessage')).toBe(data.props.errorMessage);
    });

    it('adds error class to field', () => {
      expect(
        data.wrapper
          .find('.ds-c-field')
          .first()
          .hasClass('ds-c-field--error')
      ).toBe(true);
    });
  });

  describe('has inversed theme', () => {
    let data;

    beforeEach(() => {
      data = render({ inversed: true });
    });

    it('passes inversed to FormLabel', () => {
      expect(data.wrapper.find('FormLabel').prop('inversed')).toBe(true);
    });

    it('adds inversed class to field', () => {
      expect(
        data.wrapper
          .find('.ds-c-field')
          .first()
          .hasClass('ds-c-field--inverse')
      ).toBe(true);
    });
  });

  describe('event handlers', () => {
    let data;

    beforeEach(() => {
      data = render({
        onBlur: jest.fn(),
        onChange: jest.fn()
      });
    });

    it('calls onBlur', () => {
      data.wrapper
        .find('.ds-c-field')
        .first()
        .simulate('blur');

      expect(data.props.onBlur.mock.calls.length).toBe(1);
    });

    it('calls onChange', () => {
      data.wrapper
        .find('.ds-c-field')
        .first()
        .simulate('change');

      expect(data.props.onChange.mock.calls.length).toBe(1);
    });
  });

  describe('masks', () => {
    it('exports unmaskValue method', () => {
      expect(typeof unmaskValue).toBe('function');
    });

    it('renders TextField', () => {
      const data = render();

      expect(data.wrapper).toMatchSnapshot();
    });

    it('renders currency mask', () => {
      const data = render({
        mask: 'currency'
      });

      expect(data.wrapper).toMatchSnapshot();
    });

    it('updates input classes when props are updated', () => {
      const wrapper = render({ mask: 'currency' }, true).wrapper;

      expect(wrapper.find('input').hasClass('ds-c-field--error')).toBe(false);

      wrapper.setProps({ errorMessage: 'Oh no' });

      expect(wrapper.find('input').hasClass('ds-c-field--error')).toBe(true);
    });
  });
});
