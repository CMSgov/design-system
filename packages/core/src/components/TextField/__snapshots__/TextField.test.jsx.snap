// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField masked renders currency mask 1`] = `
<div
  className="ds-u-clearfix"
>
  <FormLabel
    component="label"
    fieldId="textfield_27"
  >
    Foo
  </FormLabel>
  <div
    className="ds-c-field-mask ds-c-field-mask--currency"
  >
    <div
      className="ds-c-field__before ds-c-field__before--currency"
    >
      $
    </div>
    <input
      aria-label="Foo. Enter amount in dollars."
      className="ds-c-field ds-c-field--currency"
      id="textfield_27"
      name="spec-field"
      type="text"
    />
  </div>
</div>
`;
